samples/kotlin-android-app-destinations/src/test/java/com/segment/analytics/destinations/plugins/FirebaseDestinationTest.kt

```diff
@@ -0,0 +1,304 @@
+package com.segment.analytics.destinations.plugins
+
+import android.app.Activity
+import android.content.Context
+import android.content.pm.ActivityInfo
+import android.content.pm.PackageManager
+import android.os.Build
+import android.os.Bundle
+import androidx.core.os.bundleOf
+import com.google.firebase.analytics.FirebaseAnalytics
+import com.segment.analytics.kotlin.core.*
+import com.segment.analytics.kotlin.core.platform.Plugin
+import com.segment.analytics.kotlin.core.platform.plugins.log
+import io.mockk.*
+import io.mockk.impl.annotations.MockK
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.buildJsonArray
+import kotlinx.serialization.json.buildJsonObject
+import kotlinx.serialization.json.put
+import org.junit.Assert.*
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.annotation.Config
+
+@RunWith(RobolectricTestRunner::class)
+@Config(sdk = [Build.VERSION_CODES.O_MR1])
+class FirebaseDestinationTests {
+
+    @MockK
+    lateinit var context: Context
+
+    @MockK(relaxUnitFun = true)
+    lateinit var mockedAnalytics: Analytics
+
+    @MockK(relaxUnitFun = true)
+    lateinit var mockedFirebase: FirebaseAnalytics
+
+    private val firebaseDestination: FirebaseDestination
+
+    init {
+        MockKAnnotations.init(this)
+        mockkStatic(FirebaseAnalytics::class)
+        every { FirebaseAnalytics.getInstance(context) } returns mockedFirebase
+
+        // Need to mock log since its used in the destination
+//        mockkStatic("com.segment.analytics.kotlin.core.platform.plugins.LoggerKt")
+        mockkStatic(Class.forName("com.segment.analytics.kotlin.core.platform.plugins.LoggerKt").kotlin)
+        every { mockedAnalytics.log(any(), any(), any()) } just Runs
+
+
+        firebaseDestination = FirebaseDestination(context)
+        firebaseDestination.analytics = mockedAnalytics
+    }
+
+    @Test
+    fun `settings are updated correctly`() {
+        // An example settings blob
+        val settingsBlob: Settings = Json.decodeFromString(
+            """
+            {
+              "integrations": {
+                "Firebase": {
+                }
+              }
+            }
+        """.trimIndent()
+        )
+        firebaseDestination.update(settingsBlob, Plugin.UpdateType.Initial)
+
+        assertNotNull(firebaseDestination.firebaseAnalytics)
+    }
+
+    @Test
+    fun `identify is handled correctly`() {
+        firebaseDestination.firebaseAnalytics = FirebaseAnalytics.getInstance(context)
```

Comment:
not sure if the following works for this case, but we can mock private properties by doing something like this
```kotlin
firebaseDestination = spyk(FirebaseDestination())
every { firebaseDestination getProperty "firebaseAnalytics" } propertyType FirebaseAnalytics::class returns mockedFirebase
```
in this case, we can keep `firebaseAnalytics` private in `FirebaseDestination`, and move the setup the `setup` method avoiding doing it for every unit test

Reply:
we definitely do not want to mock / spy on the main testing object i.e. `firebaseDestination`. 

The only reason i am doing this here is cos technically that initialization during the `update` and i dont want to call update everytime i do a test, so this is more like a test setup thing. Also this needs to happen for all but settings tests... Maybe there is a better way to do this 🤔

Reply:
yeah. I agree we should not mock the main testing object. but I think spy in some cases are fine, like for this case, it just injects a mocked property. other behaviors are still the actual behaviors. but you're also right, that's kind risky if we're not paying attention. I think the current way is fine, it's just a bit strange that we makes properties `internal` just because we need them in unit tests.  😅